engine/consensus/matching/engine.go

```diff
@@ -293,16 +290,24 @@ func (e *Engine) onApproval(originID flow.Identifier, approval *flow.ResultAppro
 		return engine.NewInvalidInputErrorf("invalid origin for approval: %x", originID)
 	}
 
+	// if the approval is for an unknown block, cache it. It will be picked up
+	// later when the finalizer processes new blocks.
+	_, err := e.state.AtBlockID(approval.Body.BlockID).Head()
+	if err != nil {
+		_, _ = e.approvals.Add(approval)
```

Comment:
⚠️ 
This is dangerous, as we do _not_ perform any correctness checks here and just add the approval to the mempool. Hence, need to account for the possibility that approvals in the mempool are arbitrarily invalid. Therefore, you would need to perform the checks https://github.com/onflow/flow-go/blob/4b7c696b01e7881db12631f7a49d6ed4123c1e54/engine/consensus/matching/engine.go#L313-L321 also in `matchChunk` (and the signature checks, which are not implemented yet).